File: javascript/packages/linter/src/rules/string-utils.ts
/**
 * Checks if parentheses in a string are balanced
 * Returns false if there are more closing parens than opening at any point
 */
export function hasBalancedParentheses(content: string): boolean {
  let depth = 0

  for (const char of content) {
    if (char === "(") depth++
    if (char === ")") depth--
    if (depth < 0) return false
  }

  return depth === 0
}

/**
 * Splits a string by commas at the top level only
 * Respects nested parentheses, brackets, braces, and strings
 *
 * @example
 * splitByTopLevelComma("a, b, c") // ["a", " b", " c"]
 * splitByTopLevelComma("a, (b, c), d") // ["a", " (b, c)", " d"]
 * splitByTopLevelComma('a, "b, c", d') // ["a", ' "b, c"', " d"]
 */
export function splitByTopLevelComma(str: string): string[] {
  const result: string[] = []

  let current = ""
  let parenDepth = 0
  let bracketDepth = 0
  let braceDepth = 0
  let inString = false
  let stringChar = ""

  for (let i = 0; i < str.length; i++) {
    const char = str[i]
    const previousChar = i > 0 ? str[i - 1] : ""

    if ((char === '"' || char === "'") && previousChar !== "\\") {
      if (!inString) {
        inString = true
        stringChar = char
      } else if (char === stringChar) {
        inString = false
      }
    }

    if (!inString) {
      if (char === "(") parenDepth++
      if (char === ")") parenDepth--
      if (char === "[") bracketDepth++
      if (char === "]") bracketDepth--
      if (char === "{") braceDepth++
      if (char === "}") braceDepth--

      if (char === "," && parenDepth === 0 && bracketDepth === 0 && braceDepth === 0) {
        result.push(current)
        current = ""
        continue
      }
    }

    current += char
  }

  if (current) {
    result.push(current)
  }

  return result
}

File: rust/herb-linter/src/rules/string_utils.rs
pub fn has_balanced_parentheses(content: &str) -> bool {
  let mut depth: i32 = 0;

  for character in content.chars() {
    if character == '(' {
      depth += 1;
    }

    if character == ')' {
      depth -= 1;
    }

    if depth < 0 {
      return false;
    }
  }

  depth == 0
}

pub fn split_by_top_level_comma(string: &str) -> Vec<&str> {
  let mut result = Vec::new();
  let mut start = 0;
  let mut paren_depth = 0i32;
  let mut bracket_depth = 0i32;
  let mut brace_depth = 0i32;
  let mut in_string = false;
  let mut string_char = ' ';
  let bytes = string.as_bytes();

  for index in 0..bytes.len() {
    let character = bytes[index] as char;
    let previous_character = if index > 0 { bytes[index - 1] as char } else { '\0' };

    if (character == '"' || character == '\'') && previous_character != '\\' {
      if !in_string {
        in_string = true;
        string_char = character;
      } else if character == string_char {
        in_string = false;
      }
    }

    if !in_string {
      match character {
        '(' => paren_depth += 1,
        ')' => paren_depth -= 1,
        '[' => bracket_depth += 1,
        ']' => bracket_depth -= 1,
        '{' => brace_depth += 1,
        '}' => brace_depth -= 1,
        ',' if paren_depth == 0 && bracket_depth == 0 && brace_depth == 0 => {
          result.push(&string[start..index]);
          start = index + 1;
          continue;
        }
        _ => {}
      }
    }
  }

  if start < string.len() {
    result.push(&string[start..]);
  }

  result
}
